File: 29Items-Projects/7Algorithms/6-Backtracking/python/permutations.py
"""Permutations — generate all orderings of a set of elements."""


class Permutations:
    def __init__(self, elements: list[int]):
        self.arr = elements[:]
        self.results: list[list[int]] = []

    def solve_and_print(self) -> None:
        print(f"Elements: {self.arr}\n")

        self._backtrack(0)

        print(f"Found {len(self.results)} permutation(s):")
        for i, perm in enumerate(self.results, 1):
            print(f"  #{i}: {perm}")

    def _backtrack(self, start: int) -> None:
        if start == len(self.arr):
            self.results.append(self.arr[:])
            return
        for i in range(start, len(self.arr)):
            self.arr[start], self.arr[i] = self.arr[i], self.arr[start]
            self._backtrack(start + 1)
            self.arr[start], self.arr[i] = self.arr[i], self.arr[start]

File: 29Items-Projects/7Algorithms/6-Backtracking/java/src/permutations/Permutations.java
package permutations;

import java.util.*;

public class Permutations {

    private final int[] arr;
    private final List<List<Integer>> results = new ArrayList<>();

    public Permutations(int[] arr) {
        this.arr = arr.clone();
    }

    public void solveAndPrint() {
        System.out.println("Elements: " + Arrays.toString(arr));
        System.out.println();

        backtrack(0);

        System.out.printf("Found %d permutation(s):%n", results.size());
        for (int i = 0; i < results.size(); i++) {
            System.out.printf("  #%d: %s%n", i + 1, results.get(i));
        }
    }

    private void backtrack(int start) {
        if (start == arr.length) {
            List<Integer> perm = new ArrayList<>();
            for (int v : arr) perm.add(v);
            results.add(perm);
            return;
        }
        for (int i = start; i < arr.length; i++) {
            swap(start, i);
            backtrack(start + 1);
            swap(start, i);
        }
    }

    private void swap(int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
}
